### PR TITLE
[Site Logo]: Add logo replace flow in Inspector controls

### LINKED ISSUE #49273: Site Logo: Make it possible to set / update / remove via the Inspector
## What problem does this address?
The Site Logo block makes it possible to set a logo or update an existing one, but only via the block toolbar. If you're mid-workflow flicking between the canvas and Inspector this can feel a little unexpected. 

## What is your proposed solution?
Make it possible to set / update / remove the site logo (and site icon) via the block Inspector.

 When setting a site logo initially, we have a new "Media" panel in the inspector:

Site logo inspector, Setting initially

The Add media button would mainly do the same as the "Upload" button in the canvas.

Once an image is set, the file and a small thumbnail is shown in the inspector:

Site logo inspector, editing existing

---

This issue was updated Apr 13.

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/49273

This PR adds the actions for set, replace and reset the site logo in Inspector controls. For users with no permissions to upload, if a logo exists it will show the 'chip' only.

## Testing Instructions
1. Insert a Site Logo and try the different flows

## Screenshots or screencast 

https://user-images.githubusercontent.com/16275880/233666637-83a35da4-1094-4f37-9360-ad5418c6ae6d.mov

### CODE REVIEW

File: packages/block-library/src/site-logo/editor.scss
Comment (ntsekouras) [inline packages/block-library/src/site-logo/editor.scss:109]: Most of these styles were copied from the `style variations` buttons. Should we consider a new 'variant' of button if we plan on reusing them? 🤔

**Comment by jasmussen** [comment]
Holy guacamole, that's impressively fast work! Looks almost ready to go. I'm just about to head out ofr the day, but I expect we can greenlight this quickly next week. Thank you.

File: packages/block-library/src/site-logo/edit.js
Comment (mcsf) [inline packages/block-library/src/site-logo/edit.js]:
```suggestion
											{ isLoading ? <Spinner /> : label }
```

File: packages/block-library/src/site-logo/edit.js
Comment (mcsf) [inline packages/block-library/src/site-logo/edit.js]: It's slightly surprising that this just works; I guess it's more of a "label element" than a name.

The destination component, `MediaReplaceFlow`, doesn't type its props in the code but it does document them in its README file. It turns out that [the name is supposed to be a string](https://github.com/WordPress/gutenberg/blob/5c1d6ba207fce7c06ad94d51a62c204ae34d82f7/packages/block-editor/src/components/media-replace-flow/README.md#name). What should we do here? Leave that component untouched and work around? Or amend `MediaReplaceFlow` to explicitly accept elements? I wonder if it's safer to add a separate prop (like `label`) or just make `name` polymorphic. I have no opinion here; maybe the latter option is enough.

**Review by mcsf — COMMENTED** [review]
Looks pretty good. :)

File: packages/block-library/src/site-logo/edit.js
Comment (ntsekouras) [reply packages/block-library/src/site-logo/edit.js]: Oh yes great catch! I forgot to add the comment here yesterday and cc @ciampo and @mirka about this. Right now is as you say a workaround, without adding new props. We could deprecate 'name' and add a new polymorphic prop to handle this with a better name(no pun intended).

**Comment by jasmussen** [comment]
Really really nice work, incredibly close right out of the gate. Thank you! This is virtually ready to land. Here's a GIF showing all the inspector cases for the new media button:

site logo

A few small things:

Screenshot 2023-04-24 at 08 48 31

This should either say "Add media" (I could see us allowing a video site logo in the future) or "Add image" if we want to postpone the more generic term.

Screenshot 2023-04-24 at 08 48 50

The button should be 40px tall (indeed we should be deprecating 36px, see #46734). There should be a prop to leverage this.

In the mockup, I have image width grouped with "Media", but I'm waffling on that one. Since it's already in the Settings panel, maybe we keep that and revisit this aspect when we can refresh the Image inspector?

Screenshot 2023-04-24 at 08 49 17

In the mockups, I imagined also seeing the file extension of the file:

Screenshot 2023-04-24 at 08 53 02

To me that's actually about indicating what that field is about, actual file management. It's not critical we get this right, especially if this is a challenge, but I think it would be a nice improvement.

The media field that holds the filename and circular thumbnail is meant to be an ItemGroup, because theoretically in the future it could become a layer management interface.

Screenshot 2023-04-24 at 08 56 11

So conceptually it's similar to other ItemGroups, like the recently landed Duotone:

Screenshot 2023-04-24 at 08 55 06

I don't know how many styles we get for free by actually using the ItemGroup here, but hopefully the metrics. The item should be 40px tall:

Screenshot 2023-04-24 at 08 54 06

The thumbnail should be 20x20:

Screenshot 2023-04-24 at 08 54 14

And it should have a small gray overlaid semitransparent border, like color swatches do:

Screenshot 2023-04-24 at 08 58 15

Hopefully using the components should net all those for "free".

Nice work. Small fixes, and IMO this is ready to land.

**Comment by jameskoster** [comment]
It might be good to update the featured image inspector UI to match this as a follow-up?

**Comment by jasmussen** [comment]
Yes, good enhancement. And Cover, honestly. Cover might need a little design.

**Comment by ntsekouras** [comment]
I think I addressed the feedback. Let me know if anything more is needed.

> In the mockup, I have image width grouped with "Media", but I'm waffling on that one. Since it's already in the Settings panel, maybe we keep that and revisit this aspect when we can refresh the Image inspector?

I agree. We can follow up on that one.

File: packages/block-library/src/site-logo/edit.js
Comment (Mamaduka) [inline packages/block-library/src/site-logo/edit.js]: Can we define this component outside of the Edit component? So we don't have to re-create it on every rerender. I believe that's the recommended practice in React.

**Comment by jasmussen** [comment]
Thank you for such fast work! This is really close, and could honestly land as is. Small note, `Add media`, not `Add Media`.

There are a couple of nitpicky questions that mostly relate to the componentry and how it's used, but it reveals itself primarily in how these hover styles present themselves. For example in the empty state, the hover state of media button is animated and makes the outline thicker, making it appear as if it's focused:

hover

For that style of button (gray-bordered), that's not the correct style compared to the style variation buttons which also have the gray border. 

Related to that, the filled in media button also has this hover style:

hover media item

But ItemGroup items do not:

itemgroup hover

Overall, we should unify on a single hover style that's the same across outline buttons an itemgroups aside, which is mainly about a blue hover style. It can also be a blue outline, like the Style variation buttons, but they should not be thicker like the focus style. 

So my main question is: where are these hover styles inherited from? Ideally we shouldn't need any custom CSS at all for these components. I'm not personally averse to it, but just noting that our components should be able to handle these things natively, so that if one day we advance the hover style of the buttons, they are advanced across all instances.

**Comment by ntsekouras** [comment]
`ItemGroup` is a component to group items and the examples shared above have different inner components. That's why the styles don't match. There seems to be the styles we override for `styles button` also use the `secondary` variant and it would include more changes in components package. I think we should handle this separately in a different PR. 

For now I updated the styles introduced here to match the `styles` button.

File: packages/block-library/src/site-logo/editor.scss
Comment (jasmussen) [reply packages/block-library/src/site-logo/editor.scss:109]: Reusing is my only remaining concern about this PR, because exatcly yes, we should have as little custom CSS as possible in PRs like these, and ideally reuse componentry whenever possible.

What would it require in this place? Do the style variation buttons need to be canonized as part of the Button style? Or is it a toggle? And if we canonize those can we reuse them here? If either answer is _yes_, it would be great to do that, either here or as a followup. Thanks so much for working on this.

Comment (ntsekouras) [reply packages/block-library/src/site-logo/editor.scss:109]: In my last commit I removed some obsolete css, so it's way less now. Aside of that, these outline buttons seem to be a combination of `secondary` variant Buttons and a few overrides. I think it would make sense to have a specific Button variant for such styles.

Noting though that these styles can be used probably in just a `div` or something and it would be good to have a common styling - not sure where it should live in code though.. An example is this PR too, where we use similar styles(not the hover ones) for users that have no permissions to update, so we render the info in a `div`.

I think we can handle this in a follow up, but 🤷 . Whatever you prefer.

**Review by jasmussen — APPROVED** [review]
Visually and on its behavior, this one is ready:

status

I shared some general notes on componentry reuse here: https://github.com/WordPress/gutenberg/pull/49992#discussion_r1176148509 — I will defer to devs on the componentry on how when and where to unify here. It's a fuzzy rule: generally we should not write custom CSS for pieces like these, as it's all meant to be reused base componentry. But I wouldn't block this PR for it. We just need to all pull in that direction so that if we were to, say, update the border style on the componentry level, all linked components would inherit the new styles, instead of having local overrides that causes drift.

**Review by Mamaduka — APPROVED** [review]
Excellent work, Nik!

File: packages/block-library/src/site-logo/edit.js
Comment (ciampo) [reply packages/block-library/src/site-logo/edit.js]: It's working because internally `name` is just interpolated in the JSX:

https://github.com/WordPress/gutenberg/blob/c6029852e1bd5af06aa87d55a6789620f7c6a727/packages/block-editor/src/components/media-replace-flow/index.js#L147-L155

Under a technical point of view, it feels safe to change the docs and allow `name` to be any `ReactNode`. I guess it depends more on whether you'd prefer to "limit" consumers of the component to pass strings of text for design / semantics / any other reasons.

**Comment by mtias** [comment]
This is a great improvement, thanks for working on it!

**Comment by paaljoachim** [comment]
I added a comment here in the closed issue: https://github.com/WordPress/gutenberg/issues/49273#issuecomment-1528967005
In regards to expanding this media control to also other blocks such as Cover and Group.

**Comment by ndiego** [comment]
@ntsekouras @jasmussen I was just testing this for 15.7 and found that if the file name for the image is really long, the image will get constrained below 20px. We may want to fix ahead of the 15.7 release on Wednesday. 

cc @bph 

image

**Pull request merged**
